src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java

```diff
@@ -0,0 +1,197 @@
+package com.atlassian.bitbucket.jenkins.internal.deployments;
+
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironmentType;
+import com.atlassian.bitbucket.jenkins.internal.provider.JenkinsProvider;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMRepository;
+import com.atlassian.bitbucket.jenkins.internal.status.BitbucketRevisionAction;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.model.AbstractProject;
+import hudson.model.Item;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Notifier;
+import hudson.tasks.Publisher;
+import hudson.util.FormValidation;
+import hudson.util.ListBoxModel;
+import jenkins.model.Jenkins;
+import jenkins.tasks.SimpleBuildStep;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.AncestorInPath;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.verb.POST;
+
+import javax.annotation.CheckForNull;
+import javax.inject.Inject;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.stripToNull;
+
+public class DeployedToEnvironmentNotifierStep extends Notifier implements SimpleBuildStep {
+
+    private static final Logger LOGGER = Logger.getLogger(DeployedToEnvironmentNotifierStep.class.getName());
+
+    private final BitbucketDeploymentEnvironment environment;
+
+    @DataBoundConstructor
+    public DeployedToEnvironmentNotifierStep(String environmentKey, String environmentName,
+                                             @CheckForNull String environmentType,
+                                             @CheckForNull String environmentUrl) {
+        environment = new BitbucketDeploymentEnvironment.Builder(getOrGenerateEnvironmentKey(environmentKey),
```

Comment:
you have a rather annoying bug here. Because you have `requireNonNull`  on the `key` and `name` it will throw an NPE, this exception will be displayed to the user with the full stack trace. Yes, they did the wrong thing, and no they should not get a stack trace thrown in their face because of it. Also because the message for the NPE is just "name" it is next to impossible to work out exactly what field it relates to. I am very well aware that some other plugins do throw exceptions in the users face on error. We need to do better than that.

@mhenschke-atl and I talked about this quickly, and essentially we have a number of bad options available to us. We both agree that current behaviour constitutes a bug, so we cannot ship with it. 

Options are
1. Provide a better message -> only marginally better, still a stack trace thrown at the user. Least preferred option
2. Accept that `null` is a thing, this is how the "trigger after build" works. This gives an acceptable user experience, and places the burden on you to validate that the Step is actually valid. Our Server config is done this way. This is not excellent but a decided improvement over the current situation
3. Do validation on save. This is more work, not quite as nice user experience as 2. but better technically (no `null` nor any invalid config being saved).

_So_ I propose we go with option 3. I think it strikes a balance between technical desires, and user experience. What you need to do is to override the `newInstance` in the descriptor like so:
```java
 @Override
        public Publisher newInstance(@Nullable StaplerRequest req, @Nonnull JSONObject formData) throws FormException {
            String envName = stripToNull(formData.getString("environmentName"));
            if (envName == null) {
                throw new FormException("Name is needed", "name");
            }
            return super.newInstance(req, formData);
        }

```
For bonus points (but totally optional) you can also handle the absence of the key, I don't think you need to though, as it is a form submit, so the key will only be missing when someone calls you directly and they can get a 500 or whatever if they send malformed requests, no user will get there by mistake.

Reply:
Sorry, I changed this to 'option 1' before I saw your comment about it. Will go with Option 3.